File: src/types/result.ts
/**
 * Evaluation Result Type Definitions
 *
 * The canonical schema for individual evaluation results. Each result
 * captures everything needed for reproducibility and auditing:
 * - Deterministic run identity
 * - Correctness assessment
 * - Output artefacts and metrics
 * - Provenance information
 */

import type { ArtefactReference, Primitive } from "./case.js";
import type { SutRole } from "./sut.js";

/**
 * Categories of evaluation failure.
 */
export type FailureType =
	| "no_output" // Algorithm produced no output
	| "invalid_structure" // Output has wrong structure
	| "constraint_violation" // Output violates constraints
	| "exception" // Algorithm threw an exception
	| "oracle_mismatch" // Output doesn't match oracle
	| "timeout"; // Algorithm exceeded time limit

/**
 * Run identity and context.
 */
export interface RunContext {
	/** Deterministic run ID (hash of inputs) */
	runId: string;

	/** SUT identifier */
	sut: string;

	/** SUT role (primary/baseline/oracle) */
	sutRole: SutRole;

	/** SUT version for reproducibility */
	sutVersion?: string;

	/** Case identifier */
	caseId: string;

	/** Case class for grouping */
	caseClass?: string;

	/** Configuration overrides for this run */
	config?: Record<string, Primitive>;

	/** Random seed if applicable */
	seed?: number;

	/** Repetition number for statistical runs */
	repetition?: number;
}

/**
 * Correctness assessment.
 */
export interface CorrectnessResult {
	/** Whether expected output exists (oracle available) */
	expectedExists: boolean;

	/** Whether the SUT produced any output */
	producedOutput: boolean;

	/** Whether output is structurally valid */
	valid: boolean;

	/** Whether output matches expected (null if no oracle) */
	matchesExpected: boolean | null;

	/** Failure classification if applicable */
	failureType?: FailureType;

	/** Human-readable failure notes */
	notes?: string[];
}

/**
 * A ranked item for ranking tasks.
 */
export interface RankedItem {
	/** Item identifier */
	itemId: string;

	/** Score or rank value */
	score: number;

	/** Optional additional metadata */
	metadata?: Record<string, Primitive>;
}

/**
 * Output artefacts and summaries.
 */
export interface ResultOutputs {
	/** Scalar summary values */
	summary?: Record<string, Primitive | Primitive[]>;

	/** Classification labels */
	labels?: Record<string, Primitive>;

	/** Ranking results */
	ranking?: RankedItem[];

	/** References to generated artefacts */
	artefacts?: ArtefactReference[];

	/** Additional untyped outputs */
	extra?: Record<string, unknown>;
}

/**
 * Numeric metrics collected during evaluation.
 */
export interface ResultMetrics {
	/** Primary numeric metrics */
	numeric: Record<string, number>;

	/** Additional metrics (overflow) */
	extra?: Record<string, number>;

	/** Allow arbitrary number properties for scenario-specific metrics */
	[key: string]: number | Record<string, number> | undefined;
}

/**
 * Provenance information for reproducibility.
 */
export interface Provenance {
	/** Execution environment */
	runtime: {
		platform: string;
		arch: string;
		nodeVersion: string;
	};

	/** Git commit hash */
	gitCommit?: string;

	/** Whether working directory had uncommitted changes */
	dirty?: boolean;

	/** Hash of package-lock.json for dependency pinning */
	dependencyLockHash?: string;

	/** Parent run IDs (for derived results) */
	parentRunIds?: string[];

	/** Execution timestamp */
	timestamp?: string;

	/** Wall-clock execution time in milliseconds */
	executionTimeMs?: number;

	/** Peak memory usage during execution (bytes) */
	peakMemoryBytes?: number;

	/** Memory usage at completion (bytes) */
	finalMemoryBytes?: number;
}

/**
 * Complete evaluation result.
 *
 * This is the canonical schema for all evaluation outputs. Every experiment
 * produces results in this format for consistent aggregation and rendering.
 */
export interface EvaluationResult {
	/** Run identity and context */
	run: RunContext;

	/** Correctness assessment */
	correctness: CorrectnessResult;

	/** Output artefacts and summaries */
	outputs: ResultOutputs;

	/** Numeric metrics */
	metrics: ResultMetrics;

	/** Provenance for reproducibility */
	provenance: Provenance;
}

/**
 * Batch of evaluation results.
 */
export interface ResultBatch {
	/** Schema version */
	version: string;

	/** Generation timestamp */
	timestamp: string;

	/** All results in this batch */
	results: EvaluationResult[];

	/** Optional batch-level metadata */
	metadata?: Record<string, Primitive>;
}

File: python/src/ppef/types/result.py
"""Evaluation result type definitions."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from .case import Primitive
from .sut import SutRole

type FailureType = Literal[
    "no_output",
    "invalid_structure",
    "constraint_violation",
    "exception",
    "oracle_mismatch",
    "timeout",
]


class RunContext(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    run_id: str = Field(alias="runId")
    sut: str
    sut_role: SutRole = Field(alias="sutRole")
    sut_version: str | None = Field(default=None, alias="sutVersion")
    case_id: str = Field(alias="caseId")
    case_class: str | None = Field(default=None, alias="caseClass")
    config: dict[str, Primitive] | None = None
    seed: int | None = None
    repetition: int | None = None


class CorrectnessResult(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    expected_exists: bool = Field(alias="expectedExists")
    produced_output: bool = Field(alias="producedOutput")
    valid: bool
    matches_expected: bool | None = Field(alias="matchesExpected")
    failure_type: FailureType | None = Field(default=None, alias="failureType")
    notes: list[str] | None = None


class RankedItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    item_id: str = Field(alias="itemId")
    score: float
    metadata: dict[str, Primitive] | None = None


class ArtefactReference(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    type: str
    uri: str
    hash: str | None = None
    metadata: dict[str, Primitive] | None = None


class ResultOutputs(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    summary: dict[str, Primitive | list[Primitive]] | None = None
    labels: dict[str, Primitive] | None = None
    ranking: list[RankedItem] | None = None
    artefacts: list[ArtefactReference] | None = None
    extra: dict[str, object] | None = None


class ResultMetrics(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    numeric: dict[str, float]
    extra: dict[str, float] | None = None


class Provenance(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    runtime: dict[str, str]
    git_commit: str | None = Field(default=None, alias="gitCommit")
    dirty: bool | None = None
    dependency_lock_hash: str | None = Field(default=None, alias="dependencyLockHash")
    parent_run_ids: list[str] | None = Field(default=None, alias="parentRunIds")
    timestamp: str | None = None
    execution_time_ms: float | None = Field(default=None, alias="executionTimeMs")
    peak_memory_bytes: int | None = Field(default=None, alias="peakMemoryBytes")
    final_memory_bytes: int | None = Field(default=None, alias="finalMemoryBytes")


class EvaluationResult(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    run: RunContext
    correctness: CorrectnessResult
    outputs: ResultOutputs
    metrics: ResultMetrics
    provenance: Provenance
    error: str | None = None


class ResultBatch(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    version: str
    timestamp: str
    results: list[EvaluationResult]
    metadata: dict[str, Primitive] | None = None
